lib/auth/ec2_join.go

```diff
@@ -0,0 +1,755 @@
+			}
+		}
+		// iid matches this allow rule. Check if it is running.
+		return trace.Wrap(checkInstanceRunning(ctx, iid.InstanceID, iid.Region, rule.AWSRole))
+	}
+	return trace.AccessDenied("instance did not match any allow rules")
+}
+
+func checkInstanceRunning(ctx context.Context, instanceID, region, IAMRole string) error {
+	awsClientConfig, err := config.LoadDefaultConfig(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	awsClientConfig.Region = region
+
+	// assume the configured IAM role if necessary
+	if IAMRole != "" {
+		stsClient := sts.NewFromConfig(awsClientConfig)
+		creds := stscreds.NewAssumeRoleProvider(stsClient, IAMRole)
+		awsClientConfig.Credentials = aws.NewCredentialsCache(creds)
+	}
+
+	ec2Client := ec2ClientFromConfig(ctx, awsClientConfig)
+
+	output, err := ec2Client.DescribeInstances(ctx, &ec2.DescribeInstancesInput{
+		InstanceIds: []string{instanceID},
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if len(output.Reservations) == 0 || len(output.Reservations[0].Instances) == 0 {
+		return trace.AccessDenied("")
+	}
+	instance := output.Reservations[0].Instances[0]
+	if instance.InstanceId == nil || *instance.InstanceId != instanceID {
+		return trace.AccessDenied("")
+	}
+	if instance.State == nil || instance.State.Name != ec2types.InstanceStateNameRunning {
+		return trace.AccessDenied("")
+	}
+	return nil
+}
+
+// parseAndVerifyIID parses the given Instance Identity Document and checks that
+// the signature is valid.
+func parseAndVerifyIID(iidBytes []byte) (*imds.InstanceIdentityDocument, error) {
+	sigPEM := fmt.Sprintf("-----BEGIN PKCS7-----\n%s\n-----END PKCS7-----", string(iidBytes))
+	sigBER, _ := pem.Decode([]byte(sigPEM))
+	if sigBER == nil {
+		return nil, trace.AccessDenied("unable to decode Instance Identity Document")
+	}
+
+	p7, err := pkcs7.Parse(sigBER.Bytes)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var iid imds.InstanceIdentityDocument
+	if err := json.Unmarshal(p7.Content, &iid); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	rawCert, ok := awsCertBytes[iid.Region]
+	if !ok {
+		return nil, trace.AccessDenied("unsupported EC2 region: %q", iid.Region)
+	}
+	certPEM, _ := pem.Decode(rawCert)
+	cert, err := x509.ParseCertificate(certPEM.Bytes)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	p7.Certificates = []*x509.Certificate{cert}
+	if err = p7.Verify(); err != nil {
+		return nil, trace.AccessDenied("invalid signature")
+	}
+
+	return &iid, nil
+}
+
+func checkPendingTime(iid *imds.InstanceIdentityDocument, clock clockwork.Clock) error {
+	if clock.Since(iid.PendingTime) > 5*time.Minute {
+		return trace.AccessDenied("Instance Identity Document with PendingTime %v is older than 5 minute TTL", iid.PendingTime)
+	}
+	return nil
+}
+
+// checkInstanceUnique makes sure this instance has not already joined the
+// cluster.
+func (a *Server) checkInstanceUnique(ctx context.Context, req RegisterUsingTokenRequest, iid *imds.InstanceIdentityDocument) error {
```

Comment:
Yeah, so there is a bit of a race here, and multiple "nodes" absolutely could join at the same time with the same IID. What I currently have here is a best-effort to catch any accidental or malicious "node" that attempts to join *after* the original, "good" node has joined.

This would be somewhat difficult to strengthen without making the feature flaky. The way this currently works is
1. Node sends register request
2. Auth responds with signed certificates
3. Node heartbeats and upserts itself into the backend.

Auth only knows about the node after the heartbeat. This is nice, because if there is any network error and the node doesn't get the certificate it can just register again. For example, if auth stored the IID at step 1 and then blocked all future register requests with that IID, the node could get locked out if it never received the certificate response.

The drawback is the race you mentioned, and the fact that a malicious "node" which joins before the good node could never heartbeat, or heartbeat with a different HostID so that it goes undetected.

This is a bit weaker than what we originally discussed when coming up with this feature @klizhentas so I wonder if you have any thoughts on if this is okay or how to strengthen it

Reply:
Can we define malicious node here? I'm assuming it's not a completely external node, but the node that is authorized to get the certificate? Can you describe this sequence in more detail?

Reply:
@klizhentas A malicious node would be an actor trying to gain host credentials for the cluster in order to do something nefarious by presenting itself as a node or somehow misusing our authenticated endpoints. Today, you need a provision token for the cluster in order to get credentials. With this feature, instead of a provision token, what a malicious actor needs is a signed EC2 Instance Identity Document (IID) which matches the configured allow rules.

An IID could be "stolen" by any process running on an EC2 instance matching the allow rules (by calling the IMDS endpoint to retrieve it directly), or if it ended up in a log somewhere, or potentially by other means.

We discussed and I've implemented here 2 protections against use of a stolen IID.  The protections are:
1. The IID must be recent. The `PendingTime` field, which is the time the EC2 instance started, must be within 5 minutes. This protects against use of old IIDs found in old logs, etc.
2. The IID cannot be used more than once. If both the "good" node and the malicious node try to join with the same IID then whichever one joins first will be allowed and the second will be rejected and a warning will be logged.

For the implementation of (2) above, we discussed deriving the node name (`HostID`) from the Instance ID so that during a register request we can check for the existence of a node with the unique `HostID` derived from the IID and deny the request if it is already present. This is vulnerable to an attack with the following flow:
1. The attacker "steals" the IID
2. The attacker executes a Register request before the "good" node from that instance has sent a heartbeat and been added to the backend collection of nodes.
3. The attacker does not heartbeat, or heartbeats with a forged `HostID`
4. When the "good" node joins, it will appear to be unique so no warning will be logged.

I could see a couple potential solutions here:
1. Atomically save the IID (or a hash) to the backend during the register request and don't allow it to be used again. We only need to save the IID until the TTL expires. Could be flaky if because there is some network error and the node doesn't get, or loses, the credentials in the response it will be locked out.
2. Create a new gRPC streaming Register endpoint which confirms that the Node has received (and persisted) the credentials before saving the IID to prevent re-use. Similar to (1) above, requires a bit more work/complexity, but should be less flaky.
3. Leave the implementation as-is. We can not prevent a malicious node from joining the cluster if it has the IID and joins before the "good" node, this weakness only prevents us from logging a warning when the "good" node later joins.

Reply:
@nklaassen Let's leave the implementation as-is. Solutions 1 and 2 are going to make the system quite fragile as they require a registry. I think recency check is enough. The nodes should be only register several minutes after the instance doc is created and instance is created, preventing reuse. When searching the nodes in GetNodes, I would add a warning event emitted if there are two nodes with the same instance IDs but with a different hostname.